arraylist.c

```diff
@@ -87,6 +87,20 @@ array_list_add(struct array_list *arr, void *data)
 	return array_list_put_idx(arr, arr->length, data);
 }
 
+/*
+ * Deleting the idx-th element in the array_list.
+ */
+void
+array_list_del_idx(struct array_list *arr, int idx)
+{
+	if (idx < 0 || idx >= arr->length) {
+		return;
+	}
+	if(arr->array[idx]) arr->free_fn(arr->array[idx]);
+	memmove(arr->array + idx, arr->array + idx + 1, (--arr->length - idx) * sizeof(void *));
```

Comment:
if removing the last element, that is, if `idx == (arr->length-1)` then `arr->array + idx + 1` will point to random memory after the array - will probably cause compiler checker errors - in this case, you do not need the memmove

Reply:
also when removing the last element, the expression `(--arr->length - idx) * sizeof(void *)` will be `0` - it is possible that given a `0` argument `memmove` is a no-op, but it isn't clear from this implementation https://code.woboq.org/userspace/glibc/string/memmove.c.html that src will never be accessed

Reply:
Hmmm, this is one of my test cases and valgrind did not complain an invalid memory access... :p
```
"mmk3":[{"a0":"v0"},{},{"a1":"v1"},{"a2":"v2"},{},{},{"a3":"v3"},{}]
==>
"mmk3": [ { "a0": "v0" }, { "a1": "v1" }, { "a2": "v2" }, { "a3": "v3" } ]
```
If idx is the last one, the size (--arr->length - idx) * sizeof(void *) would be 0.  In the case, memmove just returns?